### PR TITLE
TitleView

### LINKED ISSUE #1716: [Enhancement] Page TitleView
## Rationale

Users can currently set a text-only title on a `Page` that is inside of a `NavigationPage`, and that title will be shown in the `NavigationBar`. Adding a `TitleView` attached property to `NavigationPage` will offer a greater amount of design flexibility, since it will allow users to put any `View` in the `NavigationBar`.

## Implementation

Add new attached bindable property to `NavigationPage`:

```csharp
public static readonly BindableProperty TitleViewProperty = BindableProperty.CreateAttached("TitleView", typeof(View), typeof(NavigationPage), null);
```

`NavigationPage` will need to be edited to inflate and display the new `TitleView`.

### Android

- If the `NavigationBar` is visible and the child `Page`'s `TitleView` is not null, the `TitleView` will be displayed.

### iOS

See Android

### UWP

See Android

## Implications for CSS

None

## Backward Compatibility

Since we are adding new properties, there should be no compatibility problems as long as:

1. We ensure that a `Page`'s TitleView defaults to null, as is the expected behavior now.
2. We ensure that any Effects or Custom Renderers that may have been created to produce this behavior take precedence over any values that we set, to the best of our ability.

Third party renderers will need to be updated to ensure that this functionality is officially supported.

## Difficulty : Medium
The biggest challenge here will likely be the now variable size of the `NavigationBar`. We will need to ensure that the height is reported properly so that the `Page` is given appropriate room to render.

See https://forums.xamarin.com/discussion/84526/titleview-for-pages for original proposal.
Related to #1688

### PR DESCRIPTION
**Pull request by samhouts**
### Description of Change ###

`TitleView`, like `TitleIcon`, is an attached property on `NavigationPage`. You can set it on any `Page`, and when that `Page` is pushed onto a `NavigationPage`, it will respect the value of the property.

Xaml:
```xaml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="Xamarin.Forms.Controls.GalleryPages.TitleView">

	<NavigationPage.TitleView>
		<StackLayout>
			<Label Text="This is my TitleView" />
			<Label Text="I can be a subtitle" />
		</StackLayout>
	</NavigationPage.TitleView>
	
	<StackLayout>
		<Label Text="Welcome to Xamarin.Forms!"
                VerticalOptions="CenterAndExpand" 
                HorizontalOptions="CenterAndExpand" />
	</StackLayout>
</ContentPage>
```

C#:
```csharp
NavigationPage.SetTitleView(this, CreateTitleView());
```

With `TitleView`, you can add any Xamarin.Forms `View` to the navigation bar of a `NavigationPage`.  However, you are limited to the standard `Height` of the navigation bar, with the exception of Android. If you would like to increase the `BarHeight` of the `NavigationPage` on Android, you may use the new Android.AppCompat platform configuration.
```csharp
using Xamarin.Forms.PlatformConfiguration;
using Xamarin.Forms.PlatformConfiguration.AndroidSpecific.AppCompat;

navPage.On<Android>().SetBarHeight(450);
```
To give the impression of an extended navigation bar, you may also try putting some of your header content in a view at the very top of your `Page`'s content and color match the navigation bar area. As an example, please see the `NavigationBarGallery SearchBarTitlePage` in this PR (https://github.com/xamarin/Xamarin.Forms/pull/2586/commits/7027da2a247736f22435fec0d11528804e8ced3e).

If you want to do the extended navigation bar on iOS, you will probably want to remove the separator line and shadow that is at the bottom of the navigation bar by default. You can use the new iOS platform configuration to do that.
```csharp
using Xamarin.Forms.PlatformConfiguration;
using Xamarin.Forms.PlatformConfiguration.iOSSpecific;

navPage.On<iOS>().SetHideNavigationBarSeparator(true);
```
Note: We did hope to implement `BarHeight` on iOS and UWP, but there are platform limitations. iOS does not support changing the height of the navigation bar at all, although Apple has indicated that if the demand is great enough, they may consider adding it as a feature in a later version. UWP's `CommandBar` does support changing the Height, but design guidelines prevent the entire content of the `CommandBar` from being shown unless the `CommandBar` is open. Thus, the extended content is clipped while the `CommandBar` is closed. UWP recommends hooking into the `Opening` and `Closing` events of the `CommandBar` to show and hide the content, but that is currently outside the scope of this feature. Some alternatives may be considered in the future.

As part of this enhancement, we have also enabled `TitleIcon` on Android `FormsAppCompatActivity` and UWP.

### Known Issues/Limitations ###

- `SearchBar` needs to be wrapped in a `Layout` with `HorizontalOptions = LayoutOptions.Fill` and `VerticalOptions = LayoutOptions.Fill` OR a specificed `WidthRequest` and `HeightRequest` to display properly on iOS.
- `TitleView` is not supported on Android `FormsApplicationActivity`. You must use Android `FormsAppCompatActivity`.
- Content may not be larger than the default size of the navigation bar on iOS/UWP. Clipping will occur.
- Since `BackTitle`, `Title`, `TitleIcon`, and `TitleView` all occupy the same space on the navigation bar, you can expect that there will be conflicts and limited space to support all of these properties at the same time. This will vary by platform and screen size. You may find that you can better achieve your desired design by using just `TitleView` instead of attempting to use a combination of properties. 

#### iOS 11+

Note that iOS TitleViews show is a different position depending on whether Large Titles is enabled.

#### Android

Note that Android supports `BarHeight`

#### UWP

capture
capture2
capture3

### Bugs Fixed ###

- Fixes #1716
- `TitleIcon` now works on Android AppCompat
- `TitleIcon` now works on UWP

### API Changes ###

Added:
- `NavigationPage.On<Android>().SetBarHeight(int);`
- `int NavigationPage.On<Android>().GetBarHeight();`
- `NavigationPage.On<iOS>().SetHideNavigationBarSeparator(bool);`
- `bool NavigationPage.On<iOS>().GetHideNavigationBarSeparator();`
- `View NavigationPage.TitleView { get; set; } //bindable property`

### Behavioral Changes ###

None

### PR Checklist ###

- [x] Has tests (if omitted, state reason in description)
- [x] Rebased on top of the target branch at time of PR
- [x] Changes adhere to coding standard
- [x] Consolidate commits as makes sense

### CODE REVIEW

File: Xamarin.Forms.Controls/GalleryPages/TitleView.xaml [hunk 1]
```diff
@@ -0,0 +1,24 @@
+﻿<?xml version="1.0" encoding="utf-8" ?>
+<ContentPage xmlns="http://xamarin.com/schemas/2014/forms"
+             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
+             x:Class="Xamarin.Forms.Controls.GalleryPages.TitleView"
+			 NavigationPage.TitleView="{StaticResource MyTitleView}">
+	
+	<ContentPage.Resources>
+		<ResourceDictionary>
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Controls/GalleryPages/TitleView.xaml]: packing views in RDs is not something we want to encourage. I fixed the sample in the PR description accordingly. no need to actually fix the sample

File: Xamarin.Forms.Core/NavigationPage.cs [hunk 2]
```diff
@@ -27,6 +27,8 @@ public class NavigationPage : Page, IPageContainer<Page>, INavigationPageControl
 
 		public static readonly BindableProperty TitleIconProperty = BindableProperty.CreateAttached("TitleIcon", typeof(FileImageSource), typeof(NavigationPage), default(FileImageSource));
 
+		public static readonly BindableProperty TitleViewProperty = BindableProperty.CreateAttached("TitleView", typeof(VisualElement), typeof(NavigationPage), null, propertyChanged: TitleViewPropertyChanged);
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NavigationPage.cs]: you mean `View` not `VisualElement`, right ? or can I pack a `Page` in there ?

File: Xamarin.Forms.Core/NavigationPage.cs [hunk 3]
```diff
@@ -125,6 +145,11 @@ public static FileImageSource GetTitleIcon(BindableObject bindable)
 			return (FileImageSource)bindable.GetValue(TitleIconProperty);
 		}
 
+		public static VisualElement GetTitleView(BindableObject bindable)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NavigationPage.cs]: `View`

File: Xamarin.Forms.Core/NavigationPage.cs [hunk 4]
```diff
@@ -222,6 +247,11 @@ public static void SetTitleIcon(BindableObject bindable, FileImageSource value)
 			bindable.SetValue(TitleIconProperty, value);
 		}
 
+		public static void SetTitleView(BindableObject bindable, VisualElement value)
```
Comment (StephaneDelcroix) [inline Xamarin.Forms.Core/NavigationPage.cs]: `View`

File: Xamarin.Forms.Core/NavigationPage.cs (hunk 2 above)
Comment (samhouts) [reply Xamarin.Forms.Core/NavigationPage.cs]: ...I had a very good reason for changing it from `View` to `VisualElement`, but I cannot for the life of me remember what it was.

File: Xamarin.Forms.Platform.UAP/ViewToRendererConverter.cs [hunk 5]
```diff
@@ -93,7 +93,7 @@ protected override Windows.Foundation.Size MeasureOverride(Windows.Foundation.Si
 					result = new Windows.Foundation.Size(request.Width, request.Height);
 				}
 
-				_view.Layout(new Rectangle(0, 0, result.Width, result.Height));
+				Layout.LayoutChildIntoBoundingRegion(_view, new Rectangle(0, 0, result.Width, result.Height));
```
Comment (samhouts) [inline Xamarin.Forms.Platform.UAP/ViewToRendererConverter.cs]: These two changes allow the view to be sized properly when there are margins on the container. Before this, the text of the TitleView was not centered properly.

Before:

before

After:

after

**Comment by xamarindevelopervietnam** [comment]
well done @samhouts

**Comment by samhouts** [comment]
build --uitests

File: Xamarin.Forms.Platform.UAP/NavigationPageRenderer.cs [hunk 6]
```diff
@@ -91,6 +94,24 @@ public string Title
...
+		public View TitleView
+		{
+			get
+			{
+				if (_currentPage == null)
+					return null;
+
+				return NavigationPage.GetTitleView(_currentPage) as View;
+			}
+			set { }
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Platform.UAP/NavigationPageRenderer.cs]: There's precedent in the file with other similar properties.

File: Xamarin.Forms.Platform.UAP/MasterDetailPageRenderer.cs [hunk 7]
```diff
@@ -221,6 +247,10 @@ void OnDetailPropertyChanged(object sender, PropertyChangedEventArgs e)
 		{
 			if (e.PropertyName == Page.TitleProperty.PropertyName || e.PropertyName == NavigationPage.CurrentPageProperty.PropertyName)
 				UpdateDetailTitle();
+			else if (e.PropertyName == NavigationPage.TitleIconProperty.PropertyName || e.PropertyName == NavigationPage.CurrentPageProperty.PropertyName)
+				UpdateDetailTitleIcon();
+			else if (e.PropertyName == NavigationPage.TitleViewProperty.PropertyName || e.PropertyName == NavigationPage.CurrentPageProperty.PropertyName)
+				UpdateDetailTitleView();
```
[the remark this answers is not in the document]
Comment (samhouts) [reply Xamarin.Forms.Platform.UAP/MasterDetailPageRenderer.cs]: good call

**Comment by rmarinho** [comment]
hey @jassmith should we retarget this to master and merge ?

**Comment by StephaneDelcroix** [comment]
build

**Comment by rmarinho** [comment]
There are 2 failures related with performance test , can we take a look at those @samhouts  ?!

**Comment by samhouts** [comment]
build --uitests

**Comment by samhouts** [comment]
Last failing UI tests are not related

**Comment by rmarinho** [comment]
Failing tests not related

**Comment by 15mgm15** [comment]
@samhouts I am curious about this, can we expect a pre release soon?

**Comment by plamenyovchev** [comment]
Have you planned a stable release soon ?

**Comment by DamienDoumer** [comment]
This seems not to appear on UWP. I'm using Xamarin.Forms 3.2.0.6347442 and here is a test code I'm trying to run. 
`<ContentPage
    x:Class="NewXFormsFeatures.Views.TextsModifications"
    xmlns="http://xamarin.com/schemas/2014/forms"
    xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml">

    <NavigationPage.TitleView>
        <StackLayout Orientation="Horizontal">
            <Label FontSize="Large" Text="Big Title" />
            <Entry WidthRequest="200" />
        </StackLayout>
    </NavigationPage.TitleView>

    <ContentPage.Content>
        <StackLayout>
            <Label
                HorizontalOptions="CenterAndExpand"
                Text="Welcome to Xamarin.Forms!"
                VerticalOptions="CenterAndExpand" />
        </StackLayout>
    </ContentPage.Content>
</ContentPage>`

But the TitleView does not show on UWP I tested it on Android and it was alright

**Comment by AArayeshi** [comment]
The binding in The TitleView not working.

**Comment by xamarindevelopervietnam** [comment]
same problem here.

**Comment by xamarindevelopervietnam** [comment]
gesture also is not working.

**Comment by AArayeshi** [comment]
It is not useful without binding and gesture.

**Comment by AArayeshi** [comment]
I have opened following issue with sample project.

https://github.com/xamarin/Xamarin.Forms/issues/3596

**Comment by alex-relov** [comment]
Hello. I'm using XF 3.2.0.839982. I want to make a shopping cart on the right side of the navigation bar. For the android application everything works, but for ios the basket does not align to the right. My code:
` 
<NavigationPage.TitleView><StackLayout Orientation="Horizontal">
            <Label x:Name="title" VerticalOptions="Center" />
            <AbsoluteLayout x:Name="Cart" VerticalOptions="Center" HorizontalOptions="EndAndExpand" Margin="0,5,10,0" />
</StackLayout>
</NavigationPage.TitleView>
`



how to solve this issue?

**Comment by AmrAlSayed0** [comment]
@alex-relov Yes, this a know issue #3679

**Comment by galadril** [comment]
Im having issues with the **Titleview on IOS**, not changing the size when the title or subtitle is changed.

So the titleview:
 ```
<NavigationPage.TitleView>
      <StackLayout
         x:Name="stackLayout"
         HorizontalOptions="FillAndExpand"
         Orientation="Vertical"
         Spacing="0"
         VerticalOptions="Center">
         <Label
            x:Name="title"
            HorizontalTextAlignment="{OnPlatform Android=Start,
                                                 iOS=Center}"
            LineBreakMode="NoWrap"
            VerticalOptions="Center"
            VerticalTextAlignment="Center" />
         <Label
            x:Name="subtitle"
            HorizontalOptions="{OnPlatform Android=StartAndExpand,
                                           iOS=CenterAndExpand}"
            HorizontalTextAlignment="{OnPlatform Android=Start,
                                                 iOS=Center}"
            IsVisible="False"
            LineBreakMode="NoWrap"
            VerticalOptions="Center"
            VerticalTextAlignment="Center" />
      </StackLayout>
   </NavigationPage.TitleView>
```

when you change the title or subtitle after loading the page, it doesn't make the titleview bigger.
See screenshots:

Initial values
afbeelding

Updating the values with longer versions
afbeelding


But when i rotate the device, it's corrected:

afbeelding



This only happens for me on iOS, and it works correctly on Android.

Any ideas and/or workarounds for this problem?

**Comment by PureWeen** [comment]
@galadril That will be fixed by https://github.com/xamarin/Xamarin.Forms/pull/3939

**Comment by galadril** [comment]
Thanks for the respons!

**Pull request merged**